design/crypto_primitives.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// https://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+//! Different MaidSafe tools need different crypto primitives. E.g. Crust only uses encryption
+//! public/private keys, Routing uses encryption and signing public/private keys, etc.
+//! Currently in some cases safe_crypto is unusable cause it doesn't expose some internal data like
+//! private keys, etc.
+//! One way to solve this and stay super flexible is to have 1-to-1 mappings with lower level
+//! crypto types from rust_sodium, miscreant, etc.
+//! This document attempts to catch safe_crypto design that would work for all our our cases.
+//!
+//! `PublicKeys` and `SecretKeys` currently contain both signing and encryption keys. In some
+//! cases, we don't need signing key at all, e.g. in Crust. So this document proposes to split
+//! `PublicKeys` and `SecretKeys` into corresponding types: `PublicEncryptKey, `SecretEncryptKey`,
+//! `PublicSignKey` and `SecretSignKey`.
+
+// no signing key
+pub struct PublicEncryptKey {
+    encrypt: box_::PublicKey,
+}
+
+impl PublicEncryptKey {
+    // No new() method, since public key should be constructed using `SecretEncryptKey`
+    // pub fn new() -> Self
+
+    /// Construct public key from bytes. Useful when it was serialized before.
+    pub fn from_bytes(public_key: &[u8]) -> Self
+
+    /// For anyone who wants to store public key.
+    pub fn as_bytes(&self) -> &[u8]
+
+    pub fn encrypt_anonymous<T: Serialize>(&self, plaintext: &T) -> Result<Vec<u8>, Error>
+
+    pub fn encrypt_anonymous_bytes(&self, plaintext: &[u8]) -> Vec<u8> {
+}
+
+/// Store secret key in `Arc` to discourage cloning it. Makes it more secure by not duplicating
+/// encryption keys in memory.
+pub struct SecretEncryptKey {
+    inner: Arc<SecretEncryptKeyInner>,
+    public: PublicKey,
+}
+
+struct SecretEncryptKeyInner {
+    encrypt: box_::SecretKey,
+}
+
+impl SecretEncryptKey {
+    /// Construct random public/private encryption key pair.
+    pub fn new() -> Self
+
+    /// Construct secret key from given bytes. Useful when secret key was serialized before.
+    // Note that we don't need to provide public key, seems like it's derived from secret key:
+    // https://docs.rs/rust_sodium/0.9.0/rust_sodium/crypto/box_/curve25519xsalsa20poly130/struct.SecretKey.html#method.public_key
+    pub fn from_bytes(secret_key: &[u8]) -> Self
```

Comment:
Does it happen for **_any_** array of bytes or does it need to be a part of valid RSA keypair ? Also is it valid for both RSA and EC algos ? If the inner working is via prime factorisation, does every number have a valid corresponding number to pair with ?

Otherwise we might want to return a result instead.

Reply:
> If the inner working is via prime factorisation, does every number have a valid corresponding number to pair with 
definitely not, will make this method return `Result` :)